curvefs/src/metaserver/space/inode_volume_space_deallocate.cpp

```diff
@@ -0,0 +1,378 @@
+                      "actually deallocate the volume space, fsId="
+                   << fsId_ << ", blockGroupOffset=" << blockGroupOffset;
+        return;
+    }
+
+    // 3
+    DeallocatableBlockGroupMap decrease;
+    auto &onedecrease = decrease[blockGroupOffset];
+    onedecrease.set_blockgroupoffset(blockGroupOffset);
+    onedecrease.mutable_decrease()->set_decreasedeallocatablesize(decreaseSize);
+    onedecrease.mutable_decrease()->mutable_inodeddeallocated()->Swap(
+        inodeUnderDeallocate);
+
+    st =
+        metaCli_->UpdateDeallocatableBlockGroup(fsId_, typicalInode, &decrease);
+    if (st != MetaStatusCode::OK) {
+        LOG(ERROR) << "InodeVolumeSpaceDeallocate update deallocatable size "
+                      "failed, fsId="
+                   << fsId_ << ", partitionId=" << partitionId_
+                   << ", blockGroupOffset=" << blockGroupOffset;
+        return;
+    }
+}
+
+// NOTE:
+// 1. update inode extent
+// 2. deallocate the corresponding space in volume
+//
+// According to this order, one is that the corresponding volume space will not
+// be deleted repeatedly, but there is a disadvantage, when the process hangs
+// up, this part of the space will not be recovered. And may need to be
+// deallocate by means of fscheck, etc.
+bool InodeVolumeSpaceDeallocate::DeallocateInode(uint64_t blockGroupOffset,
+                                                 const Uint64Vec &inodelist,
+                                                 uint64_t *decrease) {
+    std::vector<Extent> deallocatableVolumeSpace;
+
+    for (const auto &inodeId : inodelist) {
+        VolumeExtentSliceList sliceList;
+        MetaStatusCode st = calOpt_.inodeStorage->GetAllVolumeExtent(
+            fsId_, inodeId, &sliceList);
+        if (st != MetaStatusCode::OK) {
+            LOG(ERROR) << "InodeVolumeSpaceDeallocate get inode extent "
+                          "failed, fsId="
+                       << fsId_ << "partitionId=" << partitionId_
+                       << ", inodeId=" << inodeId;
+            return false;
+        }
+        VLOG(6) << "InodeVolumeSpaceDeallocate will update inode:" << inodeId
+                << ", blockGroupOffset:" << blockGroupOffset
+                << ", sliceList:" << sliceList.DebugString();
+
+        // 1
+        UpdateDeallocateInodeExtentSlice(blockGroupOffset, decrease, &sliceList,
+                                         &deallocatableVolumeSpace);
+        auto closure = new UpdateVolumeExtentClosure(nullptr, true);
+        metaCli_->AsyncUpdateVolumeExtent(fsId_, inodeId, sliceList, closure);
+        if (CURVEFS_ERROR::OK != closure->Wait()) {
+            LOG(ERROR) << "InodeVolumeSpaceDeallocate update inode:" << inodeId
+                       << ", blockGroupOffset:" << blockGroupOffset
+                       << ", to sliceList:" << sliceList.DebugString()
+                       << " failed";
+            return false;
```

Comment:
it's still leaked in this case, BTW, you don't have to allocate closure on heap

Reply:
fix